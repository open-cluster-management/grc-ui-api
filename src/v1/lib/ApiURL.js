/* Copyright (c) 2020 Red Hat, Inc. */

const ApiURL = {
  hostUrl: 'http://0.0.0.0/kubernetes/apis',
  clusterRegistryApiURL: '/apis/clusterregistry.k8s.io/v1alpha1/namespaces/',
  acmApiURL: 'open-cluster-management.io/v1',
  acmPoliciesRootURL: 'policies.open-cluster-management.io',
  acmPoliciesApiURL: '/apis/policies.open-cluster-management.io/v1/',
  acmAppsApiURL: '/apis/apps.open-cluster-management.io/v1/',
  // temporarily api url here, will update them later
  mcmComplianceApiURL: '/apis/compliance.mcm.ibm.com/v1alpha1/namespaces/',
  mcmNSApiURL: '/apis/mcm.ibm.com/v1alpha1/namespaces/',
};

export default ApiURL;
