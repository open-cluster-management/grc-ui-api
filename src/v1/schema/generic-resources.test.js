/** *****************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 ****************************************************************************** */
/* Copyright (c) 2020 Red Hat, Inc. */

import supertest from 'supertest';
import nock from 'nock';
import server, { GRAPHQL_PATH } from '../index';
import {
  kubeGetMock, mockAPIResourceList,
  mockCreateResourcesResponse, mockUpdateResourcesResponse,
} from '../mocks/GenericResources';
import ApiURL from '../lib/ApiURL';

describe('Generic Resources Resolver', () => {
  beforeAll(() => {
    // specify the url to be intercepted
    const APIServer = nock('http://0.0.0.0/kubernetes');

    // define the method to be intercepted
    APIServer.get('/').reply(200, kubeGetMock);
    APIServer.get('/apis/policy.mcm.ibm.com/v1alpha1').reply(200, mockAPIResourceList);
    APIServer.post(`${ApiURL.mcmPolicyApiURL}mcm/policies`)
      .reply(200, mockCreateResourcesResponse);
    APIServer.put(`${ApiURL.mcmComplianceApiURL}mcm/compliances/test-policy`)
      .reply(200, mockUpdateResourcesResponse);
  });

  test('Correctly Resolves Create Resources Mutation', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          createResources(
            resources: [
              {
                apiVersion: "policy.mcm.ibm.com/v1alpha1",
                kind: "Policy",
                metadata: {
                  name: "test-policy",
                  namespace: "mcm",
                  annotations: {
                  }
                },
                spec: {
                  complianceType: "musthave",
                  remediationAction: "inform",
                  namespaces: {
                    exclude: [
                      "kube-*"
                    ],
                    include: [
                      "default"
                    ]
                  },
                },
              }
            ]
          )
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  // object-templates: [
  //   {
  //     complianceType: "musthave",
  //     objectDefinition: {
  //       apiVersion: "v1",
  //       kind: "LimitRange",
  //       metadata: {
  //         name: "mem-limit-range"
  //       },
  //       spec: {
  //         limits: [
  //           {
  //             default: {
  //               memory: "512Mi"
  //             },
  //             defaultRequest: {
  //               memory: "256Mi"
  //             },
  //             type: "Container"
  //           }
  //         ]
  //       }
  //     }
  //   }
  // ]

  test('Correctly Resolves Update Resource Mutation', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          updateResource(
            name: "test-policy", 
            namespace: "mcm", 
            resourceType: "HCMCompliance", 
            body: [{
              apiVersion: "policy.mcm.ibm.com/v1alpha1", 
              kind: "Policy", 
              metadata: {
                name: "test-policy",
                finalizers: [
                  "propagator.finalizer.mcm.ibm.com"
                ],
                generation: 7,
                namespace: "mcm",
                resourceVersion: "1234567"
              },
            }]
          )
        }
      `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });

  test('Correctly Resolves Update Resource Labels Mutation', (done) => {
    supertest(server)
      .post(GRAPHQL_PATH)
      .send({
        query: `
        mutation {
          updateResourceLabels(
            name: "cluster1", 
            namespace: "cluster1", 
            resourceType: "HCMCluster", 
            body: [{
              op: "replace", 
              path: "/metadata/labels",
              value: {
                  cloud: "IBM",
                  datacenter: "toronto",
                  environment: "Dev"
                }
              }]
            )
          }
       `,
      })
      .end((err, res) => {
        expect(JSON.parse(res.text)).toMatchSnapshot();
        done();
      });
  });
});
