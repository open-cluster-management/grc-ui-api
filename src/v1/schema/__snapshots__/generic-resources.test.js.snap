// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic Resources Resolver Correctly Resolves Create Resources Mutation 1`] = `
Object {
  "data": Object {
    "createResources": Object {
      "errors": Array [],
      "result": Array [
        Object {
          "data": Object {
            "createResources": Object {
              "errors": Array [],
              "result": Array [
                Object {
                  "apiVersion": "policy.open-cluster-management.io/v1",
                  "kind": "Policy",
                  "metadata": Object {
                    "creationTimestamp": "2019-07-24T16:13:38Z",
                    "generation": 1,
                    "name": "test-policy",
                    "namespace": "mcm",
                    "resourceVersion": "1688805",
                    "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/mcm/policies/test-policy",
                    "uid": "fec7f06c-ae2d-11e9-8a41-005056a061f1",
                  },
                  "spec": Object {
                    "complianceType": "musthave",
                    "namespaces": Object {
                      "exclude": Array [
                        "kube-*",
                      ],
                      "include": Array [
                        "default",
                      ],
                    },
                    "remediationAction": "inform",
                  },
                },
              ],
            },
          },
        },
      ],
    },
  },
}
`;

exports[`Generic Resources Resolver Correctly Resolves Create and Update Resources Mutation 1`] = `
Object {
  "data": Object {
    "createAndUpdateResources": Object {
      "create": Object {
        "errors": Array [],
        "result": Array [
          Object {
            "kind": "Policy",
            "response": Object {
              "data": Object {
                "createResources": Object {
                  "errors": Array [],
                  "result": Array [
                    Object {
                      "apiVersion": "policy.open-cluster-management.io/v1",
                      "kind": "Policy",
                      "metadata": Object {
                        "creationTimestamp": "2019-07-24T16:13:38Z",
                        "generation": 1,
                        "name": "test-policy",
                        "namespace": "mcm",
                        "resourceVersion": "1688805",
                        "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/mcm/policies/test-policy",
                        "uid": "fec7f06c-ae2d-11e9-8a41-005056a061f1",
                      },
                      "spec": Object {
                        "complianceType": "musthave",
                        "namespaces": Object {
                          "exclude": Array [
                            "kube-*",
                          ],
                          "include": Array [
                            "default",
                          ],
                        },
                        "remediationAction": "inform",
                      },
                    },
                  ],
                },
              },
            },
          },
        ],
      },
      "update": Object {
        "errors": Array [],
        "result": Array [
          Object {
            "kind": "Policy",
            "response": Object {
              "data": Object {
                "updateResource": Object {
                  "apiVersion": "policy.open-cluster-management.io/v1",
                  "kind": "Policy",
                  "metadata": Object {
                    "annotations": Object {
                      "policy.open-cluster-management.io/categories": "hello",
                      "policy.open-cluster-management.io/controls": "",
                      "policy.open-cluster-management.io/standards": "",
                      "seed-generation": "1",
                    },
                    "creationTimestamp": "2019-07-24T13:27:59Z",
                    "finalizers": Array [
                      "propagator.finalizer.mcm.ibm.com",
                    ],
                    "generation": 6,
                    "name": "test-policy",
                    "namespace": "mcm",
                    "resourceVersion": "1673886",
                    "selfLink": "/apis/policy.open-cluster-management.io/v1/namespaces/mcm/policies/test-policy",
                    "uid": "da8f9553-ae16-11e9-8a41-005056a061f1",
                  },
                  "spec": Object {
                    "complianceType": "musthave",
                    "namespaces": Object {
                      "exclude": Array [
                        "kube-*",
                      ],
                      "include": Array [
                        "default",
                      ],
                    },
                    "object-templates": Array [
                      Object {
                        "complianceType": "musthave",
                        "objectDefinition": Object {
                          "apiVersion": "v1",
                          "kind": "LimitRange",
                          "metadata": Object {
                            "name": "mem-limit-range",
                          },
                          "spec": Object {
                            "limits": Array [
                              Object {
                                "default": Object {
                                  "memory": "512Mi",
                                },
                                "defaultRequest": Object {
                                  "memory": "256Mi",
                                },
                                "type": "Container",
                              },
                            ],
                          },
                        },
                      },
                    ],
                    "remediationAction": "inform",
                  },
                },
              },
            },
          },
        ],
      },
    },
  },
}
`;

exports[`Generic Resources Resolver Correctly Resolves Get Resource Locally 1`] = `
Object {
  "data": Object {
    "getResource": Object {
      "apiVersion": "v1",
      "kind": "Pod",
      "metadata": Object {
        "annotations": Object {
          "k8s.v1.cni.cncf.io/networks-status": "[{
    \\"name\\": \\"openshift-sdn\\",
    \\"interface\\": \\"eth0\\",
    \\"ips\\": [
        \\"10.128.2.73\\"
    ],
    \\"dns\\": {},
    \\"default-route\\": [
        \\"10.128.2.1\\"
    ]
}]",
          "openshift.io/scc": "restricted",
        },
        "creationTimestamp": "2020-07-23T18:13:32Z",
        "generateName": "grc-f2e12-grcui-6b756dfc76-",
        "labels": Object {
          "app": "grc",
          "app.kubernetes.io/instance": "grc-f2e12",
          "app.kubernetes.io/managed-by": "Helm",
          "app.kubernetes.io/name": "grc",
          "chart": "grc-2.0.0",
          "component": "ocm-grcui",
          "helm.sh/chart": "grc-2.0.0",
          "heritage": "Helm",
          "pod-template-hash": "6b756dfc76",
          "release": "grc-f2e12",
        },
        "name": "grc-f2e12-grcui-6b756dfc76-4fk7c",
        "namespace": "open-cluster-management",
        "ownerReferences": Array [
          Object {
            "apiVersion": "apps/v1",
            "blockOwnerDeletion": true,
            "controller": true,
            "kind": "ReplicaSet",
            "name": "grc-f2e12-grcui-6b756dfc76",
            "uid": "b402f0fc-a75f-4b18-8241-6def0abf3c90",
          },
        ],
        "resourceVersion": "9895924",
        "selfLink": "/api/v1/namespaces/open-cluster-management/pods/grc-f2e12-grcui-6b756dfc76-4fk7c",
        "uid": "0a25f3bc-993b-4079-954b-e50cd5027bf5",
      },
      "spec": Object {
        "affinity": Object {
          "nodeAffinity": Object {
            "requiredDuringSchedulingIgnoredDuringExecution": Object {
              "nodeSelectorTerms": Array [
                Object {
                  "matchExpressions": Array [
                    Object {
                      "key": "beta.kubernetes.io/arch",
                      "operator": "In",
                      "values": Array [
                        "amd64",
                        "ppc64le",
                        "s390x",
                      ],
                    },
                  ],
                },
              ],
            },
          },
          "podAntiAffinity": Object {
            "preferredDuringSchedulingIgnoredDuringExecution": Array [
              Object {
                "podAffinityTerm": Object {
                  "labelSelector": Object {
                    "matchExpressions": Array [
                      Object {
                        "key": "app",
                        "operator": "In",
                        "values": Array [
                          "grc-ui",
                        ],
                      },
                    ],
                  },
                  "topologyKey": "failure-domain.beta.kubernetes.io/zone",
                },
                "weight": 70,
              },
              Object {
                "podAffinityTerm": Object {
                  "labelSelector": Object {
                    "matchExpressions": Array [
                      Object {
                        "key": "app",
                        "operator": "In",
                        "values": Array [
                          "grc-ui",
                        ],
                      },
                    ],
                  },
                  "topologyKey": "kubernetes.io/hostname",
                },
                "weight": 35,
              },
            ],
          },
        },
        "containers": Array [
          Object {
            "env": Array [
              Object {
                "name": "grcUiApiUrl",
                "value": "https://grc-f2e12-grcuiapi:4000/grcuiapi",
              },
              Object {
                "name": "searchApiUrl",
                "value": "https://search-search-api:4010/searchapi",
              },
            ],
            "image": "quay.io/open-cluster-management/grc-ui@sha256:f0db7b0409c694f92f024cd16026172071804bf33ab58f789fd0506bc4b974fb",
            "imagePullPolicy": "Always",
            "livenessProbe": Object {
              "failureThreshold": 3,
              "httpGet": Object {
                "path": "/livenessProbe",
                "port": 3000,
                "scheme": "HTTP",
              },
              "initialDelaySeconds": 30,
              "periodSeconds": 30,
              "successThreshold": 1,
              "timeoutSeconds": 1,
            },
            "name": "grc-ui",
            "ports": Array [
              Object {
                "containerPort": 3000,
                "protocol": "TCP",
              },
            ],
            "readinessProbe": Object {
              "failureThreshold": 3,
              "httpGet": Object {
                "path": "/readinessProbe",
                "port": 3000,
                "scheme": "HTTP",
              },
              "initialDelaySeconds": 5,
              "periodSeconds": 10,
              "successThreshold": 1,
              "timeoutSeconds": 1,
            },
            "resources": Object {
              "requests": Object {
                "cpu": "200m",
                "memory": "256Mi",
              },
            },
            "securityContext": Object {
              "allowPrivilegeEscalation": false,
              "capabilities": Object {
                "drop": Array [
                  "ALL",
                  "KILL",
                  "MKNOD",
                  "SETGID",
                  "SETUID",
                ],
              },
              "privileged": false,
              "readOnlyRootFilesystem": true,
              "runAsUser": 1000640000,
            },
            "terminationMessagePath": "/dev/termination-log",
            "terminationMessagePolicy": "File",
            "volumeMounts": Array [
              Object {
                "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount",
                "name": "grc-sa-token-588p4",
                "readOnly": true,
              },
            ],
          },
        ],
        "dnsPolicy": "ClusterFirst",
        "enableServiceLinks": true,
        "imagePullSecrets": Array [
          Object {
            "name": "multiclusterhub-operator-pull-secret",
          },
        ],
        "nodeName": "ip-10-0-156-232.us-west-1.compute.internal",
        "priority": 0,
        "restartPolicy": "Always",
        "schedulerName": "default-scheduler",
        "securityContext": Object {
          "fsGroup": 1000640000,
          "runAsNonRoot": true,
          "seLinuxOptions": Object {
            "level": "s0:c25,c20",
          },
        },
        "serviceAccount": "grc-sa",
        "serviceAccountName": "grc-sa",
        "terminationGracePeriodSeconds": 30,
        "tolerations": Array [
          Object {
            "effect": "NoSchedule",
            "key": "dedicated",
            "operator": "Exists",
          },
          Object {
            "effect": "NoExecute",
            "key": "node.kubernetes.io/not-ready",
            "operator": "Exists",
            "tolerationSeconds": 300,
          },
          Object {
            "effect": "NoExecute",
            "key": "node.kubernetes.io/unreachable",
            "operator": "Exists",
            "tolerationSeconds": 300,
          },
          Object {
            "effect": "NoSchedule",
            "key": "node.kubernetes.io/memory-pressure",
            "operator": "Exists",
          },
        ],
        "volumes": Array [
          Object {
            "name": "grc-sa-token-588p4",
            "secret": Object {
              "defaultMode": 420,
              "secretName": "grc-sa-token-588p4",
            },
          },
        ],
      },
      "status": Object {
        "conditions": Array [
          Object {
            "lastProbeTime": null,
            "lastTransitionTime": "2020-07-23T18:13:32Z",
            "status": "True",
            "type": "Initialized",
          },
          Object {
            "lastProbeTime": null,
            "lastTransitionTime": "2020-07-23T18:14:17Z",
            "status": "True",
            "type": "Ready",
          },
          Object {
            "lastProbeTime": null,
            "lastTransitionTime": "2020-07-23T18:14:17Z",
            "status": "True",
            "type": "ContainersReady",
          },
          Object {
            "lastProbeTime": null,
            "lastTransitionTime": "2020-07-23T18:13:32Z",
            "status": "True",
            "type": "PodScheduled",
          },
        ],
        "containerStatuses": Array [
          Object {
            "containerID": "cri-o://baabe151ec104ea82dccd3613c6c9e25c008cd5998cbd77e6e63d8b887a53620",
            "image": "quay.io/open-cluster-management/grc-ui@sha256:f0db7b0409c694f92f024cd16026172071804bf33ab58f789fd0506bc4b974fb",
            "imageID": "quay.io/open-cluster-management/grc-ui@sha256:f0db7b0409c694f92f024cd16026172071804bf33ab58f789fd0506bc4b974fb",
            "lastState": Object {},
            "name": "grc-ui",
            "ready": true,
            "restartCount": 0,
            "started": true,
            "state": Object {
              "running": Object {
                "startedAt": "2020-07-23T18:14:09Z",
              },
            },
          },
        ],
        "hostIP": "10.0.156.232",
        "phase": "Running",
        "podIP": "10.128.2.73",
        "podIPs": Array [
          Object {
            "ip": "10.128.2.73",
          },
        ],
        "qosClass": "Burstable",
        "startTime": "2020-07-23T18:13:32Z",
      },
    },
  },
}
`;

exports[`Generic Resources Resolver Correctly Resolves Update Resource Labels Mutation 1`] = `
Object {
  "data": Object {
    "updateResourceLabels": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": Object {
          "attempts": 5,
          "status": 404,
          "statusCode": 404,
        },
      },
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "Nock: No match for request {
  \\"method\\": \\"PATCH\\",
  \\"url\\": \\"http://0.0.0.0/kubernetes/apis/internal.open-cluster-management.io/v1beta1/namespaces/cluster1/managedclusterinfos/cluster1\\",
  \\"headers\\": {
    \\"authorization\\": \\"localdev\\",
    \\"content-type\\": \\"application/json-patch+json\\",
    \\"host\\": \\"0.0.0.0\\",
    \\"accept\\": \\"application/json\\",
    \\"content-length\\": 138
  },
  \\"body\\": \\"[{\\\\\\"op\\\\\\":\\\\\\"replace\\\\\\",\\\\\\"value\\\\\\":[{\\\\\\"op\\\\\\":\\\\\\"replace\\\\\\",\\\\\\"path\\\\\\":\\\\\\"/metadata/labels\\\\\\",\\\\\\"value\\\\\\":{\\\\\\"cloud\\\\\\":\\\\\\"IBM\\\\\\",\\\\\\"datacenter\\\\\\":\\\\\\"toronto\\\\\\",\\\\\\"environment\\\\\\":\\\\\\"Dev\\\\\\"}}]}]\\"
}",
      "path": Array [
        "updateResourceLabels",
      ],
    },
  ],
}
`;

exports[`Generic Resources Resolver Correctly Resolves Update Resource Mutation 1`] = `
Object {
  "data": Object {
    "updateResource": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
      },
      "locations": Array [
        Object {
          "column": 11,
          "line": 3,
        },
      ],
      "message": "ACM ERROR cannot find matched resource type",
      "path": Array [
        "updateResource",
      ],
    },
  ],
}
`;
